core/src/apps/bitcoin/scripts_decred.py

```diff
@@ -0,0 +1,168 @@
+
+
+def input_derive_script(
+    script_type: EnumTypeInputScriptType,
+    multisig: Optional[MultisigRedeemScriptType],
+    coin: CoinInfo,
+    hash_type: int,
+    pubkey: bytes,
+    signature: bytes,
+) -> bytes:
+    if script_type == InputScriptType.SPENDADDRESS:
+        # p2pkh or p2sh
+        return scripts.input_script_p2pkh_or_p2sh(pubkey, signature, hash_type)
+    elif script_type == InputScriptType.SPENDMULTISIG:
+        # p2sh multisig
+        assert multisig is not None  # checked in sanitize_tx_input
+        signature_index = multisig_pubkey_index(multisig, pubkey)
+        return input_script_multisig(
+            multisig, signature, signature_index, hash_type, coin
+        )
+    else:
+        raise wire.ProcessError("Invalid script type")
+
+
+def input_script_multisig(
+    multisig: MultisigRedeemScriptType,
+    signature: bytes,
+    signature_index: int,
+    hash_type: int,
+    coin: CoinInfo,
+) -> bytearray:
+    signatures = multisig.signatures  # other signatures
+    if len(signatures[signature_index]) > 0:
+        raise wire.DataError("Invalid multisig parameters")
+    signatures[signature_index] = signature  # our signature
+
+    # length of the redeem script
+    pubkeys = multisig_get_pubkeys(multisig)
+    redeem_script_length = scripts.output_script_multisig_length(pubkeys, multisig.m)
+
+    # length of the result
+    total_length = 0
+    for s in signatures:
+        total_length += 1 + len(s) + 1  # length, signature, hash_type
+    total_length += 1 + redeem_script_length  # length, script
+
+    w = empty_bytearray(total_length)
+
+    for s in signatures:
+        if len(s):
+            scripts.append_signature(w, s, hash_type)
+
+    # redeem script
+    write_op_push(w, redeem_script_length)
+    scripts.write_output_script_multisig(w, pubkeys, multisig.m)
+
+    return w
+
+
+# A submission for an address hash.
+def output_script_sstxsubmissionpkh(addr: str) -> bytearray:
+    try:
+        raw_address = base58.decode_check(addr, blake256d_32)
+    except ValueError:
+        raise wire.DataError("Invalid address")
+    w = empty_bytearray(26)
+    w.append(0xBA)  # OP_SSTX
+    w.append(0x76)  # OP_DUP
+    w.append(0xA9)  # OP_HASH160
+    w.append(0x14)  # OP_DATA_20
+    w.extend(raw_address[2:])
+    w.append(0x88)  # OP_EQUALVERIFY
+    w.append(0xAC)  # OP_CHECKSIG
+    return w
+
+
+# A currently unused stake change script. Output amount has been checked to be
+# zero. The addr is also checked as to whether it pays to a zeroed hash.
+def output_script_sstxchange(addr: str) -> bytearray:
+    try:
+        raw_address = base58.decode_check(addr, blake256d_32)
+    except ValueError:
+        raise wire.DataError("Invalid address")
+    # Using change addresses is no longer common practice. Inputs are split
+    # beforehand.
+    for b in raw_address[2:]:
+        if b != 0:
+            raise wire.DataError("Only zeroed addresses accepted for sstx change")
```

Comment:
it seems pointless to even accept the `addr` argument. Let's instead generate the script statically.

Reply:
this seems to be the last open point.

it relates to the previous point about accepting the address in the change `TxOutput`, and it seems that using a constant `OUTPUT_SCRIPT_SSTXCHANGE` would actually be possible, despite address string being different for the same script